### PR TITLE
Allow config worker url in ServiceWorkerManager

### LINKED ISSUE #1158: service worker named with hash makes it hard to use in downstream.
## Description

The service worker file is named at build-time by Webpack and it contains the hash of the file content. For our use case of Voici, we still use the worker provided by Jupyterlite but the worker file name is still computed by the Voici build process. This leads to a mismatch between the file name provided by Webpack and the file provided by Jupyterlite.

## Expected behavior

Can we remove the hash from the worker file name?

## Context

- JupyterLite version:
- Operating System and version:
- Browser and version:

<details><summary>Browser Output</summary>
<pre>
Paste the output from your browser Javascript console here.
</pre>
</details>

### PR DESCRIPTION
**Pull request by trungleduc**
## References

Closes #1158 

## Code changes

Allow users to provide service worker URL in `ServiceWorkerManager` constructor

## User-facing changes

N/A

## Backwards-incompatible changes

The service worker file name has been changed. In `0.1.0`, it was `service-worker-[hash].js` with the `hash` computed by webpack, in `0.2.0` the hash is removed and the new file name is `service-worker.js`.

### CODE REVIEW

**Comment by martinRenou** [comment]
I *think* that does not fix the case for jupyterlite-xeus-python, or does it?

**Comment by martinRenou** [comment]
I understand it only fixes the Voici case, as Voici would be able to provide its own Service Worker with a specific name.

But in the case of jupyterlite-xeus-python, the issue is that it builds against a specific built version of jupyterlite, with a specific hash name for the service worker. And whenever a jupyterlite release comes in, jupyterlite-xeus-python now has an invalid information about the service worker name. Until we build again jupyterlite-xeus-python.

So I think it would be better to remove the hash entirely.

**Comment by martinRenou** [comment]
I don't recall the details exactly, maybe @jtpio does?

**Comment by trungleduc** [comment]
I removed the content hash, but I think there is no harm to allow configuring the worker URL, so I keep the initial changes.

**Comment by jtpio** [comment]
> Out of curiosity, `"@jupyterlite/server` is marked with `"bundled": false` in `jupyterlite-xeus-python`, I don't understand why there still is a worker file in the static assets of this extension.

hmm good question. Maybe digging in the git history will help.

File: packages/server/src/tokens.ts [hunk 1]
```diff
@@ -34,4 +34,16 @@ export interface IServiceWorkerManager {
...
 
+/**
+ * A namespace for `ServiceWorkerManager` class.
+ */
+export namespace IServiceWorkerManager {
+  /**
+   * An options object for initializing a worker manager.
+   */
+  export interface IOptions {
+    workerUrl?: string;
```
Comment (jtpio) [inline packages/server/src/tokens.ts:45]: Maybe we can add a docstring for `workerUrl` for consistency?

**Comment by jtpio** [comment]
It's a bit flaky sometimes. Seems to be passing after restarting it.

**Comment by trungleduc** [comment]
Thanks!

File: .gitignore [hunk 2]
```diff
@@ -128,6 +128,7 @@ app/test/data
 app/kernels
 app/extensions
 app/service-worker-*.js
```
Comment (martinRenou) [inline .gitignore:130]: Do we still need this rule?

**Review by martinRenou — COMMENTED** [review]
Just a minor comment, otherwise 🚀 !

File: .gitignore (hunk 2 above)
Comment (trungleduc) [reply .gitignore:130]: Nope

**Comment by jtpio** [comment]
Also maybe we can log this change in the migration guide under a new `Service Worker` section, since this affects downstreams: https://jupyterlite.readthedocs.io/en/latest/migration.html#to-0-2-0

**Review by jtpio — APPROVED** [review]
Thanks!

Following up with a release so you can test it downstream @trungleduc @martinRenou.

**Comment by trungleduc** [comment]
> Thanks!
> 
> Following up with a release so you can test it downstream @trungleduc @martinRenou.

Great! Thank you very much.

File: packages/server/src/service-manager.ts [hunk 3]
```diff
@@ -6,8 +6,10 @@ import { PageConfig, URLExt } from '@jupyterlab/coreutils';
 import { IServiceWorkerManager, WORKER_NAME } from './tokens';
 
 export class ServiceWorkerManager implements IServiceWorkerManager {
-  constructor() {
-    void this.initialize().catch(console.warn);
+  constructor(options?: IServiceWorkerManager.IOptions) {
+    const workerUrl =
+      options?.workerUrl ?? URLExt.join(PageConfig.getBaseUrl(), WORKER_NAME);
```
Comment (stevejpurves) [inline packages/server/src/service-manager.ts:11]: @trungleduc @jtpio is there a way to set this option? from what I can see here: https://github.com/jupyterlite/jupyterlite/blob/bde12e5209c37e278c5b84af9ebe40fc17696475/packages/server/src/app.ts#L65-L66

there's no way to pass this in when creating a new instance of the JupyterLiteServer, and no way to load options onto the page

**Comment by stevejpurves** [comment]
Thanks for these changes @trungleduc @martinRenou @jtpio 🙏🏼 the fact that the service worker filename is stable alone, means we can override this in thebe at build time. having control over the workUrl would be nice though -- but I don't think there's a way to do that yet. is it appropriate for that to be an option directly on the `JupyterLiteServer` constructor?

File: packages/server/src/service-manager.ts (hunk 3 above)
Comment (trungleduc) [reply packages/server/src/service-manager.ts:11]: you can disable the `"@jupyterlite/server-extension:service-worker"` plugin and provide your own `IServiceWorkerManager`
https://github.com/jupyterlite/jupyterlite/blob/bde12e5209c37e278c5b84af9ebe40fc17696475/packages/server-extension/src/index.ts#L200-L207

Comment (stevejpurves) [reply packages/server/src/service-manager.ts:11]: cool - I take it that you mean adding my own plugin with this `id: '@jupyterlite/server-extension:service-worker'` will automatically override / disable the default plugin, and of course replace it with the new one?

**Pull request merged**